### PR TITLE
vmware_guest: module rewrite to add many features & fixes

### PR DESCRIPTION
**Pull request by nerzhul**
##### ISSUE TYPE
 - Feature Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME
vmware_guest

##### ANSIBLE VERSION
```
ansible 2.2.0.0
```

##### SUMMARY
   
* New feature: Add template_flag attribute to define if the destination machine is a template
* New feature: create VM without using templates
* New feature: multiple NIC
* New feature: multiple disks
* New feature: custom SCSI controller types (default: paravirtual)
* New feature: NIC can now be E1000 or VMXNet3 (default)
* New feature: customize NIC mac address
* New feature: new disk option autoselect_datastore permit to select the less used datastore. If datastore field is provided, filter the datastore list before selection
* New feature: Implement disk resizing + addition when state=present and VM exists
* New feature: when state=present and vm exists, modify the current CPU, Memory and disk space
* New feature: Add guest_id attribute to customize VM guest ID in VMWare
* New feature: affect VM to a resource pool
* New feature: change VM configuration without recreating it (CPU, memory, disks, network, guest ID, resource pool)
* New feature: suspended state (thanks @berney)
* Add 'gatherfacts' state to gather facts on a VM instead of previous 'present' state ('present' ensure the VM configuration)
* Add PyVmomiCache class to cache read only object
* Various python code fixes
* Various documentation fixes
* esxi_hostname & cluster are now exclusive
* Drop ips attribute & set ip directly into networks
* Little performance fixes by removing some duplicate calls to VMWare API
* Python 3 portability fixes
* Create many functions to make the code maintainable
* Cleanup some useless attributes
* Add helper class to create:
  * SCSI controller
  * Disks
  * Network devices

### CODE REVIEW

**Comment by nerzhul** [comment]
i don't know why shippable failed

**Comment by mattclay** [comment]
@nerzhul Fix the merge conflict and I'll re-run the tests on Shippable.

**Comment by nerzhul** [comment]
@mattclay it's now rebased :)

**Comment by mattclay** [comment]
@nerzhul: Module validation failed:

```
2016-12-14 10:25:32 ============================================================================
2016-12-14 10:25:32 lib/ansible/modules/cloud/vmware/vmware_guest.py
2016-12-14 10:25:32 ============================================================================
2016-12-14 10:25:32 TRACE:
2016-12-14 10:25:32     mapping values are not allowed here
2016-12-14 10:25:32       in "vmware_guest.DOCUMENTATION", line 90, column 49:
2016-12-14 10:25:32          ... pe: Valid value is thin (default: None)
2016-12-14 10:25:32                                              ^
2016-12-14 10:25:32 ERROR: DOCUMENTATION is not valid YAML. Line 90 column 49
```

**Comment by nerzhul** [comment]
@mattclay right, i will fix YAML syntax soon for documentation

**Comment by nerzhul** [comment]
@mattclay all right now :)

**Comment by jctanner** [comment]
@dav1x can you review this please?

**Comment by nerzhul** [comment]
it seems there is a documentation string issue, i will fix it tomorrow, and i will finish the network idempotency tomorrow too

**Comment by nerzhul** [comment]
All is now good, the only remaining point to fix is the network modification detection, flag is already set to true when reconfiguring a VM with network

**Comment by aperigault** [comment]
I tested without set disk params and I get an error:
`line 923, in configure_disks\n    for expected_disk_spec in self.params.get('disk'):\nTypeError: 'NoneType' object is not iterable\n", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false`

**Comment by nerzhul** [comment]
Thanks for the report i will redo a full case testing on monday

**Comment by nerzhul** [comment]
@aperigault it seems you are not up to date, line 923 is not configure_disks.

**Comment by nerzhul** [comment]
Last commit improve module performance by removing some unused code. It seems append on the folder cache is very slow, this removal divide module exec time by 2, but maybe it should be improved another time, after merge or not

**Comment by nerzhul** [comment]
Added resource pool support

Resource pool location is also idempotent

**Comment by nerzhul** [comment]
@jctanner can you review these huge changes please ?

**Comment by aperigault** [comment]
@nerzhul line do not match because I add my needed features before test your branch. Have you test successfully without set disk param ?

**Comment by nerzhul** [comment]
@aperigault i looked at the current code and there is no null reference call on the part you talked about. If no disk is given the code part is ignored

**Comment by nerzhul** [comment]
I rebased but build failed due to another module which was added today...

```
./lib/ansible/modules/network/ldap_entry.py:        for name, value in self.module.params['attributes'].iteritems():
```

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 1]
```diff
@@ -58,25 +64,48 @@
         required: False
    template:
         description:
-            - Name of the template to deploy, if needed to create the guest (state=present).
+            - Template used to create guest.
+            - If this value is not set, VM is created without using a template.
             - If the guest exists already this setting will be ignored.
         required: False
+   template_flag:
```
Comment (bcoca) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: is_template might be a better name

**Comment by nerzhul** [comment]
@bcoca fixed

**Comment by angystardust** [comment]
@nerzhul Your enhancements are really nice. 
It would be great to have an `autoselect_esxi_hostname` just like `autoselect_datastore` to select the less used esxi host. It would be very very useful in vsphere setups in which you can't enable DRS because of license restrictions.

**Comment by aperigault** [comment]
Next error:
`Failed to find a resource group for my_esxi.prod`
I will add the fix in my PR

**Comment by nerzhul** [comment]
@angystardust yes i will add this feature, as atm cluster mode only choose first host, i added a TODO to select the proper host, (less used, first, last).

@aperigault maybe having an example of your usage case can help to understand how to debug it. Why not merge our PRs ?

**Comment by nerzhul** [comment]
What about moving CPU, memory and SCSI type directly to variable root instead of using hardware namespace ? I think it's better to have all those variables on the first level

**Comment by aperigault** [comment]
@nerzhul I have no opinion on the matter. If you prefer to have all those variables on the first level i'm ok. @jctanner No objection?

**Comment by nerzhul** [comment]
a great pro for this is a better documentation in readthedocs, and also a simpler code, more controled by AnsibleModule

**Pull request merged**